File: python/sol/setup_mpi.py
# -*- coding: utf-8 -*-
"""
setup_mpi.py
"""

import numpy as np

def alloc_buffer(Parm, Npx, Npy, Npz, iMin, iMax, jMin, jMax, kMin, kMax):

    SendBuf_x = SendBuf_y = SendBuf_z = RecvBuf_x = RecvBuf_y = RecvBuf_z = None
    f_dtype = Parm['f_dtype']

    if Npx > 1:
        nbuf = (jMax - jMin + 1) * (kMax - kMin + 1)
        SendBuf_x = np.zeros(nbuf, f_dtype)
        RecvBuf_x = np.zeros(nbuf, f_dtype)

    if Npy > 1:
        nbuf = (kMax - kMin + 1) * (iMax - iMin + 1)
        SendBuf_y = np.zeros(nbuf, f_dtype)
        RecvBuf_y = np.zeros(nbuf, f_dtype)

    if Npz > 1:
        nbuf = (iMax - iMin + 1) * (jMax - jMin + 1)
        SendBuf_z = np.zeros(nbuf, f_dtype)
        RecvBuf_z = np.zeros(nbuf, f_dtype)

    return SendBuf_x, SendBuf_y, SendBuf_z, RecvBuf_x, RecvBuf_y, RecvBuf_z

File: mpi/setup_mpi.c
/*
setup_mpi.c (MPI)
*/

#include "ost.h"

void setup_mpi(void)
{
	size_t size;

	// alloc MPI buffer
	if (Npx > 1) {
		size = (jMax - jMin + 1) * (kMax - kMin + 1) * sizeof(real_t);
		SendBuf_x = (real_t *)malloc(size);
		RecvBuf_x = (real_t *)malloc(size);
	}
	if (Npy > 1) {
		size = (kMax - kMin + 1) * (iMax - iMin + 1) * sizeof(real_t);
		SendBuf_y = (real_t *)malloc(size);
		RecvBuf_y = (real_t *)malloc(size);
	}
	if (Npz > 1) {
		size = (iMax - iMin + 1) * (jMax - jMin + 1) * sizeof(real_t);
		SendBuf_z = (real_t *)malloc(size);
		RecvBuf_z = (real_t *)malloc(size);
	}
}
